Add `Find` to `Routes` interface

joeriddles opened the pull request:
### Background
I am working on adding a middleware for [OpenTelemetry](https://opentelemetry.io) to a Chi-based service that needs to send a metric before the request has executed. We need to include a string for the API endpoint, like the route pattern, in the metric. This is currently not possible due to the way request route patterns are resolved using Chi's trie-based router, as noted by the maintainer in this issue https://github.com/go-chi/chi/issues/692#issuecomment-1004267972.

### Description
This PR proposes adding a `Find` function to the `Routes` interface. This `Find` method closely mirrors the `Match` function that already exists. Unlike `Match`, which returns a boolean value if a route exists for the path/HTTP method combination, `Find` returns the pattern that was matched.

Unit tests have been added for `Find`, as well as new tests added for `Match`. I refactored `Match` to use `Find` under the hood, since they are both doing almost the same thing.

This PR also adds a new middleware, `FindPattern`, which allows the user to pass a callback function that will receive the fully resolved route pattern from the HTTP request.

### Examples
A basic example of using `Find` in a middleware before the request is handled can be found here: https://github.com/joeriddles/chi-api-middleware, specifically this function https://github.com/joeriddles/chi-api-middleware/blob/main/main.go#L37.

A smaller example has also been added to the `_examples` folder.

Pipello commented:
That's cool feature 💯

pkieltyka commented:
hi @joeriddles what will be the official middleware repo for your package?

pkieltyka commented on _examples/api-middleware/main.go:
```diff
@@ -0,0 +1,35 @@
+package main
```
can you rename this file to, `_examples/find-pattern/main.go` ?

pkieltyka commented on _examples/api-middleware/main.go:
```diff
@@ -0,0 +1,35 @@
+}
+
+// Middleware that prints the API pattern before the request is handled
+func ApiMiddleware(router chi.Router) func(http.Handler) http.Handler {
```
can you rename this middleware to `FindPatternExample`

pkieltyka commented on _examples/api-middleware/main.go:
```diff
@@ -0,0 +1,35 @@
+			rctx := chi.NewRouteContext()
+			path := r.URL.Path
+			op := r.Method
+			api := router.Find(rctx, op, path)
```
please return "api" variable to "pattern"

joeriddles commented:
Thanks for the review @pkieltyka! I have renamed the file and variables you requested in this commit: https://github.com/go-chi/chi/pull/872/commits/5cf9bc354372b1222ce690e7638e3bff464cd4cf

> hi @joeriddles what will be the official middleware repo for your package?

Do you mean I should add a new middleware file, like `middlewares/find.go`? I can add something like this that allows the user to pass a function that should be called with the found route pattern.

joeriddles commented:
@pkieltyka I have added a new middleware here: https://github.com/go-chi/chi/pull/872/commits/f6c1df81e3d69f6a08e24f7d3ed20e3642afffb3. I also updated the example I added to use the new `FindPattern` middleware. Thanks again for reviewing!

casoetan commented:
👀 🙌🏾

batazor commented:
any news?

joeriddles commented:
@pkieltyka any chance this gets merged soon?

dils2k commented:
I am going to fork it and merge this PR in the fork  :eyes:

VojtechVitek reviewed (approved):
Hi,

The `.Find()` method on Mux is indeed a very nice contribution to chi. LGTM. Thank you!

---

But for now, I removed the `middleware.FindRoute()` in [1daabdc](https://github.com/go-chi/chi/pull/872/commits/1daabdc3893bea98415d6d95b631dd5ad4dd2b2c), as I think it needs some more thoughts / discussion.

```go
r.Use(middleware.FindPattern(r, func(pattern string) {
     fmt.Printf("pattern=%s\n", pattern)
}))
```

I think it'd make sense to rename this to `middleware.RoutePattern()` but also make the `*http.Request` and `http.ResponseWriter` available to the callback, so users could check `r.Method`/`r.Host`/... and for example respond with `HTTP 403` if they were to implement ACL inside of this callback.

If you have good ideas how to finish this new middleware API, please submit an issue with a proposal first.

---

Thank you for your contribution!

dils2k commented:
> I am going to fork the repo and merge this PR in the fork 👀

Should have waited a couple of hours 🥲

joeriddles commented:
Thanks for merging @VojtechVitek!

dils2k commented:
@joeriddles Is it expected behaviour? It seems to truncate the the actual result:
```go
m := chi.NewMux()
m.Route("/yo", func(r chi.Router) {
	r.Get("/sup", func(w http.ResponseWriter, r *http.Request) {
		w.Write([]byte("sup"))
	})
})

fmt.Println(m.Find(chi.NewRouteContext(), http.MethodGet, "/yo/sup")) // "/sup", but expected "/yo/sup"
```

VojtechVitek commented:
@joeriddles @dils2k 

btw: I have just found out that we already have this `rctx.RoutePattern()` method on `RouteContext`:

```
func Instrument(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		next.ServeHTTP(w, r)
		routePattern := chi.RouteContext(r.Context()).RoutePattern()
		measure(w, r, routePattern)
	})
}
```

Isn't it what we were after?

joeriddles commented:
> @joeriddles Is it expected behaviour? It seems to truncate the actual path:
> 
> ```go
> m := chi.NewMux()
> m.Route("/yo", func(r chi.Router) {
> 	r.Get("/sup", func(w http.ResponseWriter, r *http.Request) {
> 		w.Write([]byte("sup"))
> 	})
> })
> 
> fmt.Println(m.Find(chi.NewRouteContext(), http.MethodGet, "/yo/sup")) // "/sup", but expected "/yo/sup"
> ```

No, that is not expected behavior. I will create a fix.

joeriddles commented:
> @joeriddles @dils2k
> 
> btw: I have just found out that we already have this `rctx.RoutePattern()` method on `RouteContext`:
> 
> Isn't it what we were after?

The goal of `Find` is to determine the matched route pattern before the request is handled, similar to the existing `Match` function.

See the issue description:

> I am working on adding a middleware for [OpenTelemetry](https://opentelemetry.io/) to a Chi-based service that needs to send a metric before the request has executed. We need to include a string for the API endpoint, like the route pattern, in the metric. This is currently not possible due to the way request route patterns are resolved using Chi's trie-based router, as noted by the maintainer in this issue https://github.com/go-chi/chi/issues/692#issuecomment-1004267972.

dils2k commented:
> @joeriddles @dils2k
> 
> btw: I have just found out that we already have this `rctx.RoutePattern()` method on `RouteContext`:
> 
> ```
> func Instrument(next http.Handler) http.Handler {
> 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
> 		next.ServeHTTP(w, r)
> 		routePattern := chi.RouteContext(r.Context()).RoutePattern()
> 		measure(w, r, routePattern)
> 	})
> }
> ```
> 
> Isn't it what we were after?

The issue with RoutePattern is that it returns the pattern only if the request successfully reaches the route [see here](https://github.com/go-chi/chi/issues/692#issuecomment-1004267972).

However, in some cases (such as collecting metrics), it's necessary to determine the route pattern before the request is fully executed. This is important because requests can fail or be declined by authentication middleware.

VojtechVitek commented:
Cool, makes sense. Thanks for the explanation 👍 

@joeriddles I'll hold off with releasing new minor version until the fix is ready. Please keep me posted.

VojtechVitek commented:
@joeriddles any updates?

This is blocking the release, so I'm tempted to revert this PR until a fix for https://github.com/go-chi/chi/pull/872#issuecomment-2358971940 is ready. Thoughts?

joeriddles commented:
> This is blocking the release, so I'm tempted to revert this PR until a fix for [#872 (comment)](https://github.com/go-chi/chi/pull/872#issuecomment-2358971940) is ready. Thoughts?

@VojtechVitek a follow-up PR is created here: https://github.com/go-chi/chi/pull/954. Thanks!

dils2k commented:
> @joeriddles any updates?
> 
> This is blocking the release, so I'm tempted to revert this PR until a fix for [#872 (comment)](https://github.com/go-chi/chi/pull/872#issuecomment-2358971940) is ready. Thoughts?

Do you have any updates on when we might expect the release?

dils2k commented:
@VojtechVitek

VojtechVitek commented:
v5.2.0 is released
